### PR TITLE
Installing homestead by default.

### LINKED ISSUE #7: Local Rancher
A local version of Rancher needs to be running in the environment. This is for a single cluster and does not include multi-cluster management.

### PR DESCRIPTION
**Pull request by mattfarina**
Homestead is the current name for local rancher. This is in
development but provides a means for us to get to the Rancher
Dashboard among other things.

Closes #7

### CODE REVIEW

File: src/k8s-engine/helm.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    })
+
+    bat.stderr.on('data', (data) => {
+      err = data.toString()
+    })
+
+    bat.on('exit', (code) => {
+      console.log('code ' + code)
+      if (code === 0) {
+        resolve(dta)
```
Comment (mook-as) [inline src/k8s-engine/helm.js:31]: Wouldn't it make more sense to `JSON.parse` it first?

File: src/k8s-engine/helm.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
```
Comment (mook-as) [inline src/k8s-engine/helm.js:65]: All the comments from `list()` apply here too.

File: src/k8s-engine/helm.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        reject('Failed to list resource: ' + err)
+      }
+    })
+  });
+}
+
+async function install(name, chart, namespace) {
+  return new Promise((resolve, reject) => {
+    if (name === undefined) {
+      reject("name required to install")
```
Comment (mook-as) [inline src/k8s-engine/helm.js:76]: Yeah, TypeScript would be nice here… (Not clear why this needs to be inside the promise, but ¯\\\_(ツ)\_/¯).

File: src/k8s-engine/homestead.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+// This function ensures that homestead is running in a cluster
+// TODO: Handle upgrading homestead
+async function ensure() {
+  // Check if cluster available
+  // Check if homestead is running
+  let out
+  try {
+    await Helm.list('cattle-system');
+  } catch (e) {
+    // Cant connect to the cluster so there is a problem.
```
Comment (mook-as) [inline src/k8s-engine/homestead.js:14]:
```suggestion
    // Can't connect to the cluster so there is a problem.
```

File: src/k8s-engine/homestead.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    // Homestead isn't installed so install it.
+    try {
+      out = await Helm.install('homestead', './resources/homestead-0.0.1.tgz', 'cattle-system');
+    } catch (e2) {
+      throw new Error(`Unable to install homestead: ${e2}`);
+    }
+  }
+
+  return out
```
Comment (mook-as) [inline src/k8s-engine/homestead.js:34]: What is the return value supposed to be? In the happy case (already installed), this function returns `undefined`…

**Review by mook-as — CHANGES_REQUESTED** [review]
Semicolon use is very inconsistent in this patch; we should probably get a linter that warns about that.

**Comment by ericpromislow** [comment]
I ran eslint on docker/ with apply-changes on, and it touched every file. The main changes were:

1. drop quotes around simple key names before colons
2. always put a semi-colon at the end of a statement
3. prefer single-quotes to double-quotes
4. Change 
```
"string1" + expn1 + "string2" + expn
```
 to 
```
`string1${expn1}string2${expn2}`
```

For a multi-developer team, I'm fine running a style-checker in CI and failing if any style-checks fail.

**Comment by mook-as** [comment]
Actually, we probably should document how to get that homestead tarball, too; it's not clear where the helm chart comes from.

**Comment by mattfarina** [comment]
> Actually, we probably should document how to get that homestead tarball, too; it's not clear where the helm chart comes from.

Homestead is something that will be handled separately from us. We don't have to worry about making it public. When it's ready for people to be using (it's at 0.0.1 right now) then Darren or someone else will make it widely available.

File: src/k8s-engine/helm.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    bat.stdout.on('data', (data) => {
+      dta = data.toString()
+    })
+
+    bat.stderr.on('data', (data) => {
+      err = data.toString()
+    })
+
+    bat.on('exit', (code) => {
+      console.log('code ' + code)
```
[the remark this answers is not in the document]
Comment (mattfarina) [reply src/k8s-engine/helm.js:29]: I think we need to rewrite and clean up the logging. It's pretty excessive right now which is good for debugging but not useful outside of that.

[the remark this answers is not in the document]
Comment (mattfarina) [reply src/k8s-engine/helm.js:29]: And this bit was some debugging code that I meant to pull out. oops

File: src/k8s-engine/helm.js (hunk 3 above)
Comment (mattfarina) [reply src/k8s-engine/helm.js:76]: It technically does not need to be in the promise. If outside of it the code would need to throw an exception or use `Promise.reject`.

File: src/k8s-engine/homestead.js (hunk 5 above)
Comment (mattfarina) [reply src/k8s-engine/homestead.js:34]: Good catch. Both `helm install` and `helm status` return a Helm Release object in JSON form. The output of `helm status` was intended to be returned.

File: src/k8s-engine/minikube.js [hunk 7]
```diff
@@ -97,18 +98,28 @@ class Minikube extends EventEmitter {
       });
 
       bat.on('exit', async function(code, sig) {
-        that.clear();
         // When nested we do not want to keep going down the rabbit hole on error
         if (code == 80 && permsMsg && !nested) {
           // TODO: perms modal
           // TODO: Handle non-macos cases. This can be changed when multiple
           // hypervisors are used.
+          that.clear();
```
[the remark this answers is not in the document]
Comment (mattfarina) [reply src/k8s-engine/minikube.js:106]: I'd not thought of using try/finally as a way to get a defer. Ha.

**Comment by mattfarina** [comment]
@mook-as this is ready for another review

File: src/k8s-engine/minikube.js [hunk 8]
```diff
@@ -97,27 +98,39 @@ class Minikube extends EventEmitter {
...
+            resolve(resp);
+            return;
+          }
+
+          // Ensure homestead is running
+          console.log("starting homestead");
+          try {
+            await Homestead.ensure();
+          } catch (e) {
+            console.log(`Error starting homestead: ${e}`)
```
Comment (mook-as) [inline src/k8s-engine/minikube.js:117]:
```suggestion
            console.log(`Error starting homestead: ${e}`);
```

File: src/k8s-engine/minikube.js [hunk 9]
```diff
@@ -97,27 +98,39 @@ class Minikube extends EventEmitter {
...
+            code = 1
+          }
+
+          // Run the callback function.
+          if (code == 0) {
```
Comment (mook-as) [inline src/k8s-engine/minikube.js:122]:
```suggestion
          if (code === 0) {
```
(I don't think that's actually necessary here, just style points.)

**Review by mook-as — APPROVED** [review]
LGTM 👍

Got really annoyed at the process spawning, and found [await-spawn](https://www.npmjs.com/package/await-spawn) — no need to use it in this PR, but I might use it in a follow-up PR if you agree that seems like a sensible dependency to pull in.

**Comment by mattfarina** [comment]
@mook-as Nice find in await-spawn

**Pull request merged**